File: src/main/java/me/ctknight/uploadmanager/UploadThread.java
/*
 * Copyright (c) 2016. All rights reserved. Lai Jiewen <alanljw12345@gmail.com>
 */

package me.ctknight.uploadmanager;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Debug;
import android.os.PowerManager;
import android.os.Process;
import android.os.SystemClock;
import android.support.annotation.NonNull;
import android.support.v4.util.DebugUtils;
import android.util.Log;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import me.ctknight.uploadmanager.util.LogUtils;
import me.ctknight.uploadmanager.util.okhttputil.CountingInputStreamMultipartBody;
import okhttp3.Call;
import okhttp3.Headers;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import static java.net.HttpURLConnection.HTTP_INTERNAL_ERROR;
import static java.net.HttpURLConnection.HTTP_UNAVAILABLE;
import static me.ctknight.uploadmanager.UploadContract.UPLOAD_STATUS.CANCELED;
import static me.ctknight.uploadmanager.UploadContract.UPLOAD_STATUS.CANNOT_RESUME;
import static me.ctknight.uploadmanager.UploadContract.UPLOAD_STATUS.HTTP_DATA_ERROR;
import static me.ctknight.uploadmanager.UploadContract.UPLOAD_STATUS.RUNNING;
import static me.ctknight.uploadmanager.UploadContract.UPLOAD_STATUS.SUCCESS;
import static me.ctknight.uploadmanager.UploadContract.UPLOAD_STATUS.UNKNOWN_ERROR;
import static me.ctknight.uploadmanager.UploadContract.UPLOAD_STATUS.WAITING_FOR_NETWORK;
import static me.ctknight.uploadmanager.UploadContract.UPLOAD_STATUS.WAITING_TO_RETRY;


public class UploadThread implements Runnable, CountingInputStreamMultipartBody.ProgressListener {

    private static final int DEFAULT_TIMEOUT = (int) (20 * 1000L);

    private static final String TAG = LogUtils.makeTag(UploadThread.class);
    private static final OkHttpClient mClient = buildClient();
    private static final Object mMonitor = new Object();
    private final Context mContext;
    private final UploadNotifier mNotifier;
    private final long mId;
    private final UploadInfo mInfo;
    // global setting
    private final UploadInfoDelta mInfoDelta;
    private Call mCall;
    // upload has started or not
    private boolean mMadeProgress = false;
    private long mLastUpdateBytes = 0;
    private long mLastUpdateTime = 0;
    //TYPE_NONE
    private int mNetworkType = -1;
    private long mSpeed;
    private long mSpeedSampleStart;
    private long mSpeedSampleBytes;

    public UploadThread(Context context, UploadNotifier notifier, UploadInfo info) {
        mContext = context.getApplicationContext();
        mNotifier = notifier;

        mId = info.mId;
        mInfo = info;
        mInfoDelta = new UploadInfoDelta(info);
    }

    public static boolean isStatusRetryable(int status) {
        switch (status) {
            case HTTP_DATA_ERROR:
            case HTTP_UNAVAILABLE:
            case HTTP_INTERNAL_ERROR:
            case RUNNING:
                return true;
        }
        return false;
    }

    @NonNull
    private static OkHttpClient buildClient() {
        OkHttpClient.Builder builder = new OkHttpClient.Builder()
                .connectTimeout(DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS)
                .readTimeout(DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS);
        return builder.build();
    }

    @Override
    public void run() {
        Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND);
        final ConnectivityManager connectivityManager = (ConnectivityManager) mContext.getSystemService(Context.CONNECTIVITY_SERVICE);

        if (UploadInfo.queryUploadStatus(mContext.getContentResolver(), mId) == SUCCESS) {
            Log.d("UploadThread", "run: " + "skipping finished item id: " + mId);
            return;
        }

        PowerManager.WakeLock wakelock = null;
        final PowerManager pm = (PowerManager) mContext.getSystemService(Context.POWER_SERVICE);

        try {
            wakelock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "UploadThread" + mId);
            wakelock.acquire();

//            Log.d("UploadThread", "run: starting");

            final NetworkInfo info = connectivityManager.getActiveNetworkInfo();
            if (info != null) {
                mNetworkType = info.getType();
            }
            if (!checkDeletedOrCanceled() || isStatusRetryable(mInfoDelta.mStatus)) {
                executeUpload();
            }
            mInfoDelta.mStatus = SUCCESS;

            if (mInfoDelta.mTotalBytes == -1) {
                mInfoDelta.mTotalBytes = mInfoDelta.mCurrentBytes;
            }

        } catch (FileNotFoundException e) {
            mInfoDelta.mErrorMsg = e.getMessage();
            Log.e(TAG, "executeUpload: ", e);
            mInfoDelta.mStatus = UploadContract.UPLOAD_STATUS.FILE_NOT_FOUND;

        } catch (UploadException e) {
            mInfoDelta.mErrorMsg = e.getMessage();

            Log.w(TAG, "run: " + "Stop uploading with "
                    + mInfoDelta.mStatus
                    + mInfoDelta.mErrorMsg);

            //only we can request retry
            if (mInfoDelta.mStatus == WAITING_TO_RETRY) {
                throw new IllegalStateException("Execution should always throw final error codes");
            }

            if (isStatusRetryable(mInfoDelta.mStatus)) {
                if (mMadeProgress) {
                    mInfoDelta.mNumFailed = 1;
                } else {
                    mInfoDelta.mNumFailed += 1;
                }

                if (mInfoDelta.mNumFailed < UploadContract.Constants.MAX_RETRIES) {
                    final NetworkInfo info = connectivityManager.getActiveNetworkInfo();
                    if (info != null && info.getType() == mNetworkType && info.isConnected()) {
                        // Underlying network is still intact, use normal backoff
                        mInfoDelta.mStatus = WAITING_TO_RETRY;
                    } else {
                        // Network changed, retry on any next available
                        mInfoDelta.mStatus = WAITING_FOR_NETWORK;
                    }

                    //we can't resume uploading, so checking ETag is unnecessary.
                } else {
                    mInfoDelta.mStatus = CANNOT_RESUME;
                    mInfoDelta.writeToDatabase();
                }
            } else {
                mInfoDelta.mStatus = CANNOT_RESUME;
                mInfoDelta.writeToDatabase();
            }

        } catch (Throwable t) {
            mInfoDelta.mStatus = UNKNOWN_ERROR;
            mInfoDelta.mErrorMsg = t.toString();

            Log.e(TAG, "Failed: " + mInfoDelta.mErrorMsg, t);
        } finally {
//            Log.d(TAG, "run: Finish with status" + mInfoDelta.mStatus);

            mNotifier.notifyUploadSpeed(mId, 0);


            if (UploadContract.isComplete(mInfoDelta.mStatus)) {
                if (mInfoDelta.mVisibility == UploadContract.VISIBILITY_STATUS.VISIBLE) {
                    mInfoDelta.mVisibility = UploadContract.VISIBILITY_STATUS.VISIBLE_COMPLETE;
                }
                mInfo.sendIntentIfRequested();
            }

            mInfoDelta.writeToDatabase();

            if (wakelock != null) {
                wakelock.release();
                wakelock = null;
            }
        }
    }

    @Override
    public void transferred(long num) {
        mMadeProgress = true;
        mInfoDelta.mCurrentBytes = num;
        try {
            updateProgress();
            if (checkDeletedOrCanceled()) {
                throw new UploadCancelException("Upload canceled");
            }

        } catch (IOException | UploadException e) {
            mCall.cancel();

            Log.e(TAG, "transferred: ", e);
        }

    }

    private void executeUpload() throws UploadException, IOException {

        URL url;

        try {
            url = new URL(mInfoDelta.mTargetUrl);
        } catch (MalformedURLException e) {
            throw new UploadException("Invalid URL");
        }

        try {
            checkConnectivity();
            uploadData(url);
        } catch (Exception e) {
            if (e instanceof FileNotFoundException) {
                throw e;
            } else {
                throw new UploadNetworkException(e);
            }
        }
    }

    private void checkConnectivity() throws UploadException {

        final UploadInfo.NetworkState networkState = mInfo.checkCanUseNetwork();
        if (networkState != UploadInfo.NetworkState.OK) {
            int status = UploadContract.UPLOAD_STATUS.WAITING_FOR_NETWORK;
            if (networkState == UploadInfo.NetworkState.CANNOT_USE_ROAMING) {
                status = UploadContract.UPLOAD_STATUS.WAITING_FOR_WIFI;
                mInfo.notifyQueryForNetwork(true);
            } else if (networkState == UploadInfo.NetworkState.NO_CONNECTION) {
                status = UploadContract.UPLOAD_STATUS.WAITING_FOR_WIFI;
                mInfo.notifyQueryForNetwork(false);
            }
            throw new UploadException(status + networkState.name());
        }
    }

    private InputStream getFileInputStream() throws IOException {
        Uri fileUri = Uri.parse(mInfo.mFileUri);
        final String scheme = fileUri.getScheme();
        try {
            switch (scheme) {
                case ContentResolver.SCHEME_CONTENT:
                    return mContext.getContentResolver().openInputStream(fileUri);
                case ContentResolver.SCHEME_FILE:
                    String path = fileUri.getPath();
                    return new FileInputStream(new File(path));
                default:
                    throw new IOException("Unsupported Uri" + mInfo.mFileUri);
            }
        } catch (IOException e) {
            Log.e("UploadThread", "getFileInputStream: ", e);
            throw e;
        }
    }

    private RequestBody buildRequestBody() throws IOException {
        CountingInputStreamMultipartBody.Builder builder = new CountingInputStreamMultipartBody.Builder()
                .setType(CountingInputStreamMultipartBody.FORM)
                .addFormDataPart("file", mInfo.mFileName,
                        CountingInputStreamMultipartBody.create(MediaType.parse(mInfo.mMimeType), getFileInputStream()));
        for (Map.Entry<String, String> cd : mInfo.getContentDisposition().entrySet()) {
            builder.addFormDataPart(cd.getKey(), cd.getValue());
        }
        RequestBody body = builder.setProgressListener(this).build();
        try {
            setTotalBytes(body.contentLength());
        } catch (IOException e) {
            Log.e(TAG, "buildRequestBody: ", e);
            setTotalBytes(-1);
        }
        return body;
    }

    private Request buildRequest(URL url) throws IOException {
        return new Request.Builder()
                .headers(Headers.of(mInfo.getHeaders()))
                .url(url)
                .post(buildRequestBody())
                .build();
    }

    private void uploadData(URL url) throws IOException {
        synchronized (mMonitor) {
            mCall = mClient.newCall(buildRequest(url));
        }
        Response response = mCall.execute();
        String responseMsg = response.body().string();
        recordResponse(responseMsg);
    }

    public void setTotalBytes(long totalBytes) {
        mInfoDelta.mTotalBytes = totalBytes;
        mInfoDelta.writeToDatabase();
    }

    private void recordResponse(String responseMsg) {
        Log.d(TAG, "executeUpload: " + responseMsg);
        mInfoDelta.mServerResponse = responseMsg;
        mInfoDelta.writeToDatabase();
    }

    private void updateProgress() throws IOException, UploadException {
        final long now = SystemClock.elapsedRealtime();
        final long currentBytes = mInfoDelta.mCurrentBytes;

        final long sampleDelta = now - mSpeedSampleStart;
        if (sampleDelta > 500) {
            final long sampleSpeed = ((currentBytes - mSpeedSampleBytes) * 1000) / sampleDelta;

            if (mSpeed == 0) {
                mSpeed = sampleSpeed;
            } else {
                mSpeed = ((mSpeed * 3) + sampleSpeed) / 4;
                //From AOSP, but why??
            }

            //only notify after a full sample window time (sampleDelta)
            if (mSpeedSampleStart != 0) {
                mNotifier.notifyUploadSpeed(mId, mSpeed);
            }
            mSpeedSampleStart = now;
            mSpeedSampleBytes = currentBytes;
        }

        final long bytesDelta = currentBytes - mLastUpdateBytes;
        final long timeDelta = now - mLastUpdateTime;

        if (bytesDelta > UploadContract.Constants.MIN_PROGRESS_STEP && timeDelta > UploadContract.Constants.MIN_PROGRESS_TIME) {
            mInfoDelta.writeToDatabaseOrThrow();
            mLastUpdateBytes = currentBytes;
            mLastUpdateTime = now;
        }

    }

    private boolean checkDeletedOrCanceled() throws UploadException {
        synchronized (mInfo) {
            return mInfo.mStatus == CANCELED || mInfo.mDeleted;
        }
    }

    private class UploadInfoDelta {
        public String mTargetUrl;
        public String mMimeType;
        public int mStatus;
        public int mNumFailed;
        public int mRetryAfter;
        public long mTotalBytes;
        public long mCurrentBytes;
        public int mVisibility;

        public String mErrorMsg;
        public String mServerResponse;

        public UploadInfoDelta(UploadInfo info) {
            mTargetUrl = info.mTargetUrl;
            mMimeType = info.mMimeType;
            mStatus = info.mStatus;
            mNumFailed = info.mNumFailed;
            mRetryAfter = info.mRetryAfter;
            mTotalBytes = info.mTotalBytes;
            mCurrentBytes = info.mCurrentBytes;
            mServerResponse = info.mServerResponse;
            mVisibility = info.mVisibility;
        }

        private ContentValues buildContentValues() {
            final ContentValues values = new ContentValues();

            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_TARGET_URL, mTargetUrl);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_MIME_TYPE, mMimeType);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_STATUS, mStatus);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_NUM_FAILED, mNumFailed);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_RETRY_AFTER, mRetryAfter);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_TOTAL_BYTES, mTotalBytes);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_CURRENT_BYTES, mCurrentBytes);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_LAST_MODIFICATION, System.currentTimeMillis());
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_ERROR_MSG, mErrorMsg);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_SERVER_RESPONSE, mServerResponse);
            values.put(UploadContract.UPLOAD_COLUMNS.COLUMN_VISIBILITY, mVisibility);
            return values;
        }

        /**
         * Blindly push update of current delta values to provider.
         */
        public void writeToDatabase() {
            mContext.getContentResolver().update(mInfo.getUploadsUri(), buildContentValues(),
                    null, null);
        }

        /**
         * Push update of current delta values to provider, asserting strongly
         * that we haven't been paused or deleted.
         */
        public void writeToDatabaseOrThrow() throws UploadException {
            if (mContext.getContentResolver().update(mInfo.getUploadsUri(),
                    buildContentValues(), UploadContract.UPLOAD_COLUMNS.COLUMN_DELETED + " == '0'", null) == 0) {
                throw new UploadException(Integer.toString(CANCELED) + "Upload deleted or missing!");
            }
        }
    }
}

File: lib/src/main/java/me/ctknight/uploadmanager/internal/UploadThread.kt
/*
 * Copyright (c) 2019. All rights reserved. Lai Jiewen <alanljw12345@gmail.com>
 */

package me.ctknight.uploadmanager.internal

import android.app.job.JobParameters
import android.content.Context
import android.database.sqlite.SQLiteClosable
import android.net.ConnectivityManager
import android.net.Uri
import android.os.ParcelFileDescriptor
import android.os.Process
import android.os.SystemClock
import android.util.Log
import androidx.core.content.getSystemService
import com.zhy.http.okhttp.request.CountingRequestBody
import me.ctknight.uploadmanager.UploadContract
import me.ctknight.uploadmanager.UploadContract.UploadStatus.*
import me.ctknight.uploadmanager.UploadDatabase
import me.ctknight.uploadmanager.UploadJobService
import me.ctknight.uploadmanager.UploadRecord
import me.ctknight.uploadmanager.util.LogUtils
import me.ctknight.uploadmanager.util.OkHttpUtils
import okhttp3.*
import okhttp3.internal.Util
import java.io.FileNotFoundException
import java.io.IOException

internal class UploadThread(
    private val mJobService: UploadJobService,
    private val mClient: OkHttpClient,
    private val mParams: JobParameters,
    @Volatile private var mInfo: UploadRecord.Impl
) : Thread(), CountingRequestBody.Listener {
  private val mContext: Context = mJobService
  private val mNotifier: UploadNotifier = UploadNotifier.getInstance(mContext)
  private val mDatabase: UploadDatabase = Database.getInstance(mContext)
  private val mSystemFacade: SystemFacade = SystemFacade.Impl.getInstance(mContext)
  private val mId: Long = mInfo._ID
  private val connectivityManager: ConnectivityManager = mContext.getSystemService()!!
  // global setting
  private lateinit var mCall: Call
  //  TODO: use this list to record unclosed fds and close them at appropriate time
  private val fdList = ArrayList<ParcelFileDescriptor>(1)
  // upload has started or not
  private var mMadeProgress = false
  private var mLastUpdateBytes: Long = 0
  private var mLastUpdateTime: Long = 0
  //TYPE_NONE
  private var mNetworkType = ConnectivityManager.TYPE_DUMMY
  private var mSpeed: Long = 0
  private var mSpeedSampleStart: Long = 0
  private var mSpeedSampleBytes: Long = 0

  @Volatile
  private var mShutdownRequested: Boolean = false

  override fun run() {
    Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND)
//    connectivityManager.registerDefaultNetworkCallback(object: )
    if (mInfo.Status == SUCCESS) {
      logDebug("run: skipping finished item id: $mId")
      return
    }
    try {
      val info = mSystemFacade.getActiveNetworkInfo()
      if (info != null) {
        mNetworkType = info.type
      }

      executeUpload()

      mInfo = mInfo.copy(Status = SUCCESS)

      if (mInfo.TotalBytes == -1L) {
        mInfo = mInfo.copy(TotalBytes = mInfo.CurrentBytes)
      }
    } catch (e: StopRequestException) {
      if (!mCall.isCanceled) {
        mCall.cancel()
      }
      mInfo = mInfo.copy(ErrorMsg = e.message, Status = e.finalStatus)

      logWarning("run: Stop uploading with Status: ${mInfo.Status}, ErrorMsg: ${mInfo.ErrorMsg}")

      // Nobody below our level should request retries, since we handle
      // failure counts at this level.
      if (mInfo.Status == WAITING_TO_RETRY) {
        throw IllegalStateException("Execution should always throw final error codes")
      }

      if (mInfo.Status.isRetryable()) {
        mInfo = if (mMadeProgress) {
          mInfo.copy(NumFailed = 1)
        } else {
          mInfo.copy(NumFailed = mInfo.NumFailed + 1)
        }

        mInfo = if (mInfo.NumFailed < UploadContract.Constants.MAX_RETRIES) {
          val info = mSystemFacade.getActiveNetworkInfo()
          if (info != null && info.type == mNetworkType && info.isConnected) {
            // Underlying network is still intact, use normal backoff
            mInfo.copy(Status = WAITING_TO_RETRY)
          } else {
            // Network changed, retry on any next available
            mInfo.copy(Status = WAITING_FOR_NETWORK)
          }
        } else {
          mInfo.copy(Status = CAN_NOT_RESUME)
        }
      } else {
        mInfo = mInfo.copy(Status = CAN_NOT_RESUME)
      }

      if (mInfo.Status == WAITING_FOR_NETWORK && mInfo.isMeteredAllowed(mContext)) {
        mInfo = mInfo.copy(Status = WAITING_FOR_WIFI)
      }

    } catch (t: Throwable) {
      mInfo = mInfo.copy(Status = UNKNOWN_ERROR, ErrorMsg = t.toString())
      logError("Failed: ${mInfo.ErrorMsg}", t)
    } finally {
      logDebug("Finished with status ${mInfo.Status}")
      mNotifier.notifyUploadSpeed(mId, 0)
      mInfo.partialUpdate(mDatabase)
      closeFds()
    }

    var needsReschedule = false
    if (mInfo.Status in listOf(WAITING_FOR_WIFI, WAITING_FOR_NETWORK, WAITING_TO_RETRY)) {
      needsReschedule = true
    }
    mJobService.jobFinishedInternal(mParams, needsReschedule)
  }

  @Throws(StopRequestException::class)
  private fun getFileDescriptor(fileUri: Uri): ParcelFileDescriptor {
    try {
      return mContext.contentResolver.openFileDescriptor(fileUri, "r")
          ?: throw FileNotFoundException("The return of openFileDescriptor($fileUri) is null")
    } catch (e: FileNotFoundException) {
      logError("getFileDescriptor: ", e)
      throw StopRequestException(FILE_NOT_FOUND, e)
    }
  }

  override fun onRequestProgress(bytesWritten: Long, contentLength: Long) {
    mMadeProgress = true
    mInfo = mInfo.copy(CurrentBytes = bytesWritten, TotalBytes = contentLength)

    updateProgress()

    if (mShutdownRequested) {
      if (!mCall.isCanceled) {
        mCall.cancel()
      }
    }
  }

  @Throws(StopRequestException::class)
  private fun executeUpload() {
    checkConnectivity()
    uploadData().use { response ->
      val responseMsg = response.body()?.string()
      recordResponse(responseMsg)
      if (!response.isSuccessful) {
        StopRequestException.throwUnhandledHttpError(response.code(), response.message())
      }
    }
  }

  @Throws(StopRequestException::class)
  private fun checkConnectivity() {
    val info = mSystemFacade.getActiveNetworkInfo()
    if (info == null || !info.isConnected) {
      throw StopRequestException(WAITING_FOR_NETWORK, "Network is not connected")
    }
    if (info.isRoaming && !mInfo.isRoamingAllowed(mContext)) {
      throw StopRequestException(WAITING_FOR_NETWORK, "Waiting for not roaming Network")
    }
    if (connectivityManager.isActiveNetworkMetered && !mInfo.isMeteredAllowed(mContext)) {
      throw StopRequestException(WAITING_FOR_NETWORK, "Waiting for un-metered Wifi")
    }
  }

  @Throws(StopRequestException::class)
  private fun buildRequestBody(): RequestBody {
    val builder = MultipartBody.Builder()
        .setType(MultipartBody.FORM)
    mInfo.Parts.forEach {
      val fileInfo = it.fileInfo
      if (fileInfo != null) {
        val fileDescriptor = getFileDescriptor(fileInfo.fileUri)
        fdList.add(fileDescriptor)
        val fileRequestBody = OkHttpUtils.createRequestFromFile(fileInfo.mimeType, fileDescriptor)
        builder.addFormDataPart(it.name, fileInfo.fileName, fileRequestBody)
      } else {
        builder.addFormDataPart(it.name, it.value!!)
      }
    }

    val realBody = builder.build()
    try {
      setTotalBytes(realBody.contentLength())
    } catch (e: IOException) {
      Log.e(TAG, "buildRequestBody: ", e)
      setTotalBytes(-1)
    }

    return CountingRequestBody(realBody, this)
  }

  private fun buildRequest(): Request {
    val builder = Request.Builder()
    val headers = mInfo.Headers
    if (headers != null) {
      builder.headers(headers)
    }
    val userAgent = mInfo.UserAgent
    if (userAgent != null) {
      builder.header("User-Agent", userAgent)
    }
    return builder
        .url(mInfo.TargeUrl)
        .post(buildRequestBody())
        .build()
  }

  @Throws(StopRequestException::class)
  private fun uploadData(): Response {
    mCall = mClient.newCall(buildRequest())
    try {
      return mCall.execute()
    } catch (e: IOException) {
      throw StopRequestException(HTTP_DATA_ERROR, e)
    } catch (e: IllegalStateException) {
      throw StopRequestException(HTTP_DATA_ERROR, e)
    }
  }

  private fun setTotalBytes(totalBytes: Long) {
    mInfo = mInfo.copy(TotalBytes = totalBytes)
    mInfo.partialUpdate(mDatabase)
  }

  private fun recordResponse(responseMsg: String?) {
    logDebug("executeUpload: $responseMsg")
    mInfo = mInfo.copy(ServerResponse = responseMsg)
    mInfo.partialUpdate(mDatabase)
  }

  private fun updateProgress() {
    val now = SystemClock.elapsedRealtime()
    val currentBytes = mInfo.CurrentBytes

    val sampleDelta = now - mSpeedSampleStart
    if (sampleDelta > 500) {
      val sampleSpeed = (currentBytes - mSpeedSampleBytes) * 1000 / sampleDelta

      mSpeed = if (mSpeed == 0L) {
        sampleSpeed
      } else {
        (mSpeed * 3 + sampleSpeed) / 4
      }

      //only notify after a full sample window time (sampleDelta)
      if (mSpeedSampleStart != 0L) {
        mNotifier.notifyUploadSpeed(mId, mSpeed)
      }
      mSpeedSampleStart = now
      mSpeedSampleBytes = currentBytes
    }

    val bytesDelta = currentBytes - mLastUpdateBytes
    val timeDelta = now - mLastUpdateTime

    if (bytesDelta > UploadContract.Constants.MIN_PROGRESS_STEP && timeDelta > UploadContract.Constants.MIN_PROGRESS_TIME) {
      mInfo.partialUpdate(mDatabase)
      mLastUpdateBytes = currentBytes
      mLastUpdateTime = now
    }
  }

  private fun closeFds() {
    fdList.forEach { Util.closeQuietly(it) }
    fdList.clear()
  }

  internal fun requestShutdown() {
    mShutdownRequested = true
  }

  private fun logDebug(msg: String) {
    Log.d(TAG, "[$mId] $msg")
  }

  private fun logWarning(msg: String) {
    Log.w(TAG, "[$mId] $msg")
  }

  private fun logError(msg: String, t: Throwable) {
    Log.e(TAG, "[$mId] $msg", t)
  }

  companion object {
    private val TAG = LogUtils.makeTag<UploadThread>()
  }
}
